src/pybind/mgr/progress/module.py

```diff
@@ -260,17 +259,49 @@ def __init__(self, *args, **kwargs):
 
     def _osd_out(self, old_map, old_dump, new_map, osd_id):
         affected_pgs = []
+        unmoved_pgs = []
         for pool in old_dump['pools']:
             pool_id = pool['pool']
             for ps in range(0, pool['pg_num']):
+                up_acting = old_map.pg_to_up_acting_osds(pool['pool'], ps)
+
+                # Was this OSD affected by the OSD going out?
+                old_osds = set(up_acting['up']) | set(up_acting['acting'])
+                was_on_out_osd = osd_id in old_osds
+                if not was_on_out_osd:
+                    continue
+
                 self.log.debug("pool_id, ps = {0}, {1}".format(
                     pool_id, ps
                 ))
-                up_acting = old_map.pg_to_up_acting_osds(pool['pool'], ps)
+
                 self.log.debug(
                     "up_acting: {0}".format(json.dumps(up_acting, indent=2)))
-                if osd_id in up_acting['up'] or osd_id in up_acting['acting']:
+
+                new_up_acting = new_map.pg_to_up_acting_osds(pool['pool'], ps)
+                new_osds = set(new_up_acting['up']) | set(new_up_acting['acting'])
+
+                # Has this OSD been assigned a new location?
+                # (it might not be if there is no suitable place to move
+                #  after an OSD failure)
+                is_relocated = len(new_osds - old_osds) > 0
```

Comment:
instead, might want to put
```py
is_relocated = osd_id in old_osds - new_osds
```

Reply:
I'm not sure I get the proposed logic.  At this point in the code, it's always true that that old_osds contains osd_id (that was checked a few lines earlier).  Given that, the check you're suggesting is really equivalent to "osd_id not in new_osds".

My intent is to only create the progress event when there are some new OSDs, i.e. there is some data to be moved.

Reply:
got it! i misunderstood your intent, thought that the progress event was emitted only if the `osd_id` was *removed* from the `old_osds`.